benchmarks/cpp/nvfuser/normalization.cpp

```diff
@@ -0,0 +1,638 @@
+  // Point-wise
+  auto x_mean = div(x_sum_bcast, num_features);
+  auto x_mean_sub = sub(input, x_mean);
+
+  auto x_mean_sub_pow = mul(x_mean_sub, x_mean_sub);
+  // Reduction
+  auto var_sum = sum(x_mean_sub_pow, reduction_axes);
+  // Broadcast
+  auto var_sum_bcast = broadcast(var_sum, broadcast_mask);
+  // Point-wise
+  auto var = div(var_sum_bcast, num_features);
+  auto var_eps = add(var, new Float(kEps));
+  auto rvar = unaryOp(UnaryOpType::Rsqrt, var_eps);
+  auto output = mul(x_mean_sub, rvar);
+  return output;
+}
+
+
+//------------------------------------------------------------------------------
+
+static void MagicScheduler_Softmax(benchmark::State& benchmark_state) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  std::vector<int64_t> input_shape{benchmark_state.range(1), benchmark_state.range(0)};
+  const int kReductionAxis = benchmark_state.range(2);
+
+  // setup fusion
+  auto input = TensorViewBuilder().ndims(input_shape.size()).dtype(DataType::Float).build();
+  fusion.addInput(input);
+  auto output = mySoftmax(&fusion, input, input_shape.size(), kReductionAxis);
+  fusion.addOutput(output);
+
+  std::vector<TensorView*> reduction_tensors;
+  std::vector<TensorView*> other_tensors;
+  analyzeFusion(&fusion, reduction_tensors, other_tensors);
+
+  // inputs
+  at::manual_seed(0);
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor at_x = at::randn(input_shape, options);
+  std::vector<c10::IValue> inputs({at_x});
+
+  // outputs
+  std::vector<at::Tensor> outputs;
+
+  auto reduction_params =
+      getNormalizationHeuristics(&fusion, inputs, reduction_tensors);
+  TORCH_CHECK(reduction_params, "Reduction schedule was not generated!");
+
+  scheduleNormalization(
+      &fusion, reduction_params.value(), reduction_tensors, other_tensors);
+
+  FusionExecutor executor;
+  executor.setMeasureKernelTimeFlag(true);
+  executor.compileFusion(&fusion);
+
+  cudaDeviceSynchronize();
+  for (auto _ : benchmark_state) {
+    outputs = executor.runFusion(
+        c10::ArrayRef<c10::IValue>(inputs), reduction_params.value().lparams);
+    benchmark_state.SetIterationTime(executor.kernelTimeMs() / 1000.0);
+    cudaDeviceSynchronize();
+  }
+}
+
+static void MagicScheduler_Softmax_Baseline(benchmark::State& benchmark_state) {
+  std::vector<int64_t> input_shape{benchmark_state.range(1), benchmark_state.range(0)};
+  const int kReductionAxis = benchmark_state.range(2);
+
+  // inputs
+  at::manual_seed(0);
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor at_x = at::randn(input_shape, options);
+
+  cudaDeviceSynchronize();
+  for (auto _ : benchmark_state) {
+    // Create
+    float kernel_time_ms_ = 0;
+    cudaEvent_t start_event = {};
+    cudaEvent_t finish_event = {};
+
+    // Setup
+    cudaEventCreate(&start_event);
+    cudaEventCreate(&finish_event);
+    cudaEventRecord(start_event);
```

Comment:
this part seems to be used in multiple benchmarks - let's factor it out. one idea is to simply have a scope - RAII-type - measurement:

```cpp
{
  CudaKernelTimer timer;
  ... code ...
    benchmark_state.SetIterationTime(timer.elapsed());
    cudaDeviceSynchronize();
}
```

or even fold the last part into the timer:

```cpp
{
  CudaKernelTimer timer(benchmark_state); // it will set iteration time automatically in the destructor
  ... code ...
}
```

Reply:
:+1: